python/lsst/resources/s3utils.py

```diff
@@ -30,17 +30,19 @@
 import functools
 import os
 import re
+import urllib.parse
 from collections.abc import Callable, Iterator
 from contextlib import contextmanager
 from http.client import HTTPException, ImproperConnectionState
 from types import ModuleType
-from typing import TYPE_CHECKING, Any, cast
+from typing import TYPE_CHECKING, Any, NamedTuple, cast
 from unittest.mock import patch
 
 from botocore.exceptions import ClientError
 from botocore.handlers import validate_bucket_name
 from deprecated.sphinx import deprecated
 from urllib3.exceptions import HTTPError, RequestError
+from urllib3.util import Url, parse_url
```

Comment:
I just used it because we already have a dependency on it and it has more useful semantics for what I was trying to do... the equivalent code for getting the username/password part and reconstructing the URL without it is a lot uglier with urllib.parse because it treats the entire netloc as a single unit.